ymer/src/main/java/com/avanza/ymer/PersistedInstanceIdRecalculationService.java

```diff
@@ -57,21 +63,31 @@ public void recalculatePersistedInstanceId(String collectionName, int numberOfIn
 		if (!indexDropped) {
 			log.info("Step 1/3\tNo index to drop");
 		}
-		AtomicInteger count = new AtomicInteger();
-		collection.findByQuery(query(where(DOCUMENT_ROUTING_KEY).exists(true)))
-				.peek(it -> {
-					int currentCount = count.incrementAndGet();
-					if (currentCount % 10_000 == 0) {
-						log.info("Step 2/3\tUpdated persisted instance id for {} documents", currentCount);
-					}
-				})
-				.forEach(document -> {
-					Object id = document.get("_id");
-					int instanceId = getInstanceId(document.get(DOCUMENT_ROUTING_KEY), numberOfInstances);
-					collection.updateById(id, Map.of(DOCUMENT_INSTANCE_ID, instanceId));
-				});
 
-		log.info("Step 2/3\tUpdated persisted instance id for {} documents", count.get());
+		log.debug("Using batch size {}", BATCH_SIZE);
+		try (Stream<List<Document>> batches = collection.findByQuery(query(where(DOCUMENT_ROUTING_KEY).exists(true)), BATCH_SIZE, DOCUMENT_ROUTING_KEY, DOCUMENT_INSTANCE_ID)) {
+			AtomicInteger count = new AtomicInteger();
+			batches.forEach(batch -> collection.bulkWrite(bulkWriter -> {
+				Map<Integer, List<Document>> updatesByInstanceId = batch.stream()
+						.collect(groupingBy(it -> getInstanceId(it.get(DOCUMENT_ROUTING_KEY), numberOfInstances)));
+
+				updatesByInstanceId.forEach((instanceId, documents) -> {
+					Set<Object> ids = documents.stream()
+							.filter(document -> !Objects.equals(instanceId, document.get(DOCUMENT_INSTANCE_ID)))
+							.map(document -> document.get("_id"))
+							.filter(Objects::nonNull)
+							.peek(it -> {
+								int currentCount = count.incrementAndGet();
```

Comment:
Since this method now filters on only objects whose `_instanceId` should be updated, `currentCount` here now refers to the number of modified objects and not the number of analysed documents as before. Do you think there should be some other progress indicator about the number of analysed documents also?

Reply:
Sure, I've added a counter for the number of analyzed documents in https://github.com/AvanzaBank/ymer/pull/62/commits/ce71003adc88b92a2c9991d6d4a4acaa03d563d7